### PR TITLE
Async validation ✔️

### PR DESCRIPTION
**Pull request by NickMitrokhin**
Added async validation
Discussion: #9319

### CODE REVIEW

File: js/integration/knockout/validation.js [hunk 1]
```diff
@@ -1,51 +1,72 @@
...
-            rule.validator = that;
+import { each, map } from "../../core/utils/iterator";
+import { extend } from "../../core/utils/extend";
+import Class from "../../core/class";
+import EventsMixin from "../../core/events_mixin";
+import ValidationEngine from "../../ui/validation_engine";
+import ko from "knockout";
+
+const koDxValidator = Class.inherit({
+    ctor(target, option) {
```
Comment (DokaRus) [inline js/integration/knockout/validation.js]:
```suggestion
    ctor(target, { name, validationRules }) {
```

File: js/integration/knockout/validation.js [hunk 2]
```diff
@@ -1,51 +1,72 @@
...
+        this.target = target;
+        this.name = option.name;
+        this.isValid = ko.observable(true);
+        this.validationError = ko.observable();
+        this.validationErrors = ko.observable();
+        this.validationStatus = ko.observable("valid");
+
```
Comment (DokaRus) [inline js/integration/knockout/validation.js:17]:
```suggestion
this._validationResult = {};
```
and then you can update L26:

```js
-  if(this._validationResult && this._validationResult.status === "pending") {
+  if(this._validationResult.status === "pending") {
```

File: js/ui/button.js [hunk 3]
```diff
@@ -306,17 +306,43 @@ var Button = Widget.inherit({
...
+            } else {
+                this._setDisabled(false);
             }
-            e.stopPropagation();
+            return result;
         });
+    },
 
+    _renderSubmitInput: function() {
+        const submitAction = this._createAction((args) => {
```
Comment (DokaRus) [inline js/ui/button.js]: Extract the logic of creating an action into a separate method

Comment (DokaRus) [inline js/ui/button.js]:
```suggestion
        const submitAction = this._createAction(({ event: e }) => {
```

File: js/ui/button.js [hunk 4]
```diff
@@ -306,17 +306,43 @@ var Button = Widget.inherit({
...
+            const e = args.event;
```
Comment (DokaRus) [inline js/ui/button.js]:
```suggestion

```

File: js/ui/button.js [hunk 5]
```diff
@@ -306,17 +306,43 @@ var Button = Widget.inherit({
...
+                const validationGroup = ValidationEngine.getGroupConfig(this._findGroup());
+                if(validationGroup) {
+                    const result = validationGroup.validate();
+                    this._validationStatus = result.status;
+                    if(this._validationStatus === "pending") {
+                        this._setDisabled(true);
+                        this._waitForValidationCompleting(result.complete);
+                    }
+                }
+            }
```
Comment (DokaRus) [inline js/ui/button.js:340]:
```suggestion
            } else if(this._validationStatus !== "pending") {
                this._setDisabled(false)
            }
```

File: js/ui/button.js [hunk 6]
```diff
@@ -306,17 +306,43 @@ var Button = Widget.inherit({
...
+            this._validationStatus !== "pending" && !this._needValidate && this._setDisabled(false);
```
Comment (DokaRus) [inline js/ui/button.js]:
```suggestion

```

**Review by vconst — COMMENTED** [review]
Very large changes in dx.all.d.ts.
Not all cases are covered by tests: 
- complete callback in validationResult if several async rules are defined
- reject deferred from validationCallback

File: js/ui/editor/editor.js [hunk 7]
```diff
@@ -203,11 +217,13 @@ var Editor = Widget.inherit({
     },
 
     _renderValidationState: function() {
-        var isValid = this.option("isValid"),
-            validationError = this.option("validationError"),
+        const isValid = this.option("isValid") && this.option("validationStatus") !== "invalid",
             validationMessageMode = this.option("validationMessageMode"),
             $element = this.$element();
-
+        let validationErrors = this.option("validationErrors");
```
Comment (DokaRus) [inline js/ui/editor/editor.js:235]:
```suggestion
        const validationError = this.option("validationError");
        const validationErrors = this.option("validationErrors") || validationError ? [validationError] : null;
```
and remove L224-L226

File: js/ui/editor/editor.js [hunk 8]
```diff
@@ -221,9 +237,18 @@ var Editor = Widget.inherit({
             this._$validationMessage = null;
         }
 
-        if(!isValid && validationError && validationError.message) {
+        let validationErrorMessage = "";
+        if(validationErrors) {
```
Comment (DokaRus) [inline js/ui/editor/editor.js]: I think we can extract method `getValidationErrorMessage(validationErrors): string`

File: js/ui/text_box/ui.text_editor.base.js [hunk 9]
```diff
@@ -737,6 +781,7 @@ var TextEditorBase = Editor.inherit({
     },
 
     _dispose: function() {
+        this._disposePendingIndicator();
```
Comment (DokaRus) [inline js/ui/text_box/ui.text_editor.base.js]: Perhaps we can move it to the `_clean` method?

File: js/ui/validation/validation_mixin.js [hunk 10]
```diff
@@ -1,5 +1,5 @@
-var ValidationMixin = {
-    _findGroup: function() {
+const ValidationMixin = {
+    _findGroup() {
         var group = this.option("validationGroup"),
```
Comment (DokaRus) [inline js/ui/validation/validation_mixin.js]: Use `let` or `const` for it

File: js/ui/validation_engine.js [hunk 11]
```diff
@@ -1,43 +1,51 @@
...
-    defaultFormattedMessage: function(value) { return messageLocalization.getFormatter("validation-" + this.NAME + "-formatted")(value); },
-
-    _isValueEmpty: function(value) {
+import Class from "../core/class";
+import { extend } from "../core/utils/extend";
+import { inArray } from "../core/utils/array";
+import { each } from "../core/utils/iterator";
+import EventsMixin from "../core/events_mixin";
+import errors from "../core/errors";
+import commonUtils from "../core/utils/common";
```
Comment (DokaRus) [inline js/ui/validation_engine.js]:
```suggestion
import { grep } from "../core/utils/common";
```

File: js/ui/validation_engine.js [hunk 12]
```diff
@@ -1,43 +1,51 @@
...
+import typeUtils from "../core/utils/type";
```
Comment (DokaRus) [inline js/ui/validation_engine.js:8]:
```suggestion
import {
    isDefined,
    isString,
    isNumeric,
    isFunction
} from "../core/utils/type";
```

File: js/ui/validation_engine.js [hunk 13]
```diff
@@ -1,43 +1,51 @@
...
+import messageLocalization from "../localization/message";
+import Promise from "../core/polyfills/promise";
+import { fromPromise, Deferred } from "../core/utils/deferred";
+
+class BaseRuleValidator {
+    constructor() {
+        this.NAME = "base";
+    }
+    defaultMessage(value) {
+        return messageLocalization.getFormatter("validation-" + this.NAME)(value);
```
Comment (DokaRus) [inline js/ui/validation_engine.js]:
```suggestion
        return messageLocalization.getFormatter(`validation-${this.NAME}`)(value);
```

File: js/ui/validation_engine.js [hunk 14]
```diff
@@ -1,43 +1,51 @@
...
+    }
+    defaultFormattedMessage(value) {
+        return messageLocalization.getFormatter("validation-" + this.NAME + "-formatted")(value);
```
Comment (DokaRus) [inline js/ui/validation_engine.js]:
```suggestion
        return messageLocalization.getFormatter(`validation-${this.NAME}-formatted`)(value);
```

File: js/ui/validation_engine.js [hunk 15]
```diff
@@ -573,95 +801,227 @@ var ValidationEngine = {
...
+                items: asyncRuleItems,
+                result,
+                name
+            });
+        }
+        result.status = result.pendingRules ? "pending" : (result.isValid ? "valid" : "invalid");
+        return result;
+    },
 
+    _validateAsyncRules(info) {
```
Comment (DokaRus) [inline js/ui/validation_engine.js]:
```suggestion
    _validateAsyncRules({ result, value, items, name }) {
```
And remove L921

File: js/ui/validation_engine.js [hunk 16]
```diff
@@ -573,95 +801,227 @@ var ValidationEngine = {
...
+                result,
+                values
+            });
+        });
         return result;
     },
 
-    registerValidatorInGroup: function(group, validator) {
-        var groupConfig = ValidationEngine.addGroup(group);
+    _updateRuleConfig(info) {
```
Comment (DokaRus) [inline js/ui/validation_engine.js]:
```suggestion
    _updateRuleConfig({ rule, ruleResult, validator, name }) {
```
And remove L961

File: js/ui/validation_engine.js [hunk 17]
```diff
@@ -573,95 +801,227 @@ var ValidationEngine = {
...
+                    validator: validator,
+                    name
+                });
+            }
+        }
+    },
 
-        if(inArray(validator, groupConfig.validators) < 0) {
-            groupConfig.validators.push(validator);
+    _getPatchedRuleResult(info) {
```
Comment (DokaRus) [inline js/ui/validation_engine.js]:
```suggestion
    _getPatchedRuleResult({ ruleResult, isValid = true }) {
```
And remove L976

File: js/ui/validation_engine.js [hunk 18]
```diff
@@ -573,95 +801,227 @@ var ValidationEngine = {
...
+                isValid: typeUtils.isBoolean(ruleResult) ? ruleResult : isValid
+            };
         }
+        return result;
     },
 
-    _shouldRemoveGroup: function(group, validatorsInGroup) {
-        var isDefaultGroup = group === undefined,
-            isValidationGroupInstance = group && group.NAME === "dxValidationGroup";
+    _getAsyncRulesResult(info) {
```
Comment (DokaRus) [inline js/ui/validation_engine.js]:
```suggestion
    _getAsyncRulesResult({ values, result }) {
```
And remove L992

File: js/ui/validation_summary.js [hunk 19]
```diff
@@ -217,53 +217,61 @@ var ValidationSummary = CollectionWidget.inherit({
...
     },
 
-    _itemValidationHandler: function(itemValidationResult) {
-        var items = this.option("items"),
-            isValid = itemValidationResult.isValid,
-            elementIndex,
-            replacementFound = false,
-            newMessage = itemValidationResult.brokenRule && itemValidationResult.brokenRule.message,
-            validator = itemValidationResult.validator;
+    _itemValidationHandler(itemValidationResult) {
```
Comment (DokaRus) [inline js/ui/validation_summary.js]:
```suggestion
    _itemValidationHandler({ isValid, validator, brokenRules }) {
```
And remove L224-L226

File: js/ui/validator.js [hunk 20]
```diff
@@ -72,118 +73,116 @@ var Validator = DOMComponent.inherit({
...
+            * @type_function_param1_field5 brokenRule:RequiredRule|NumericRule|RangeRule|StringLengthRule|CustomRule|CompareRule|PatternRule|EmailRule|AsyncRule
+            * @type_function_param1_field6 brokenRules:Array<RequiredRule,NumericRule,RangeRule,StringLengthRule,CustomRule,CompareRule,PatternRule,EmailRule,AsyncRule>
+            * @type_function_param1_field6 status:Enums.ValidationStatus
             * @action
             */
 
             /**
             * @name dxValidatorOptions.rtlEnabled
             * @hidden
+            * @inheritdoc
```
Comment (DokaRus) [inline js/ui/validator.js]: Is it neccessary?

File: js/ui/validator.js [hunk 21]
```diff
@@ -219,66 +220,86 @@ var Validator = DOMComponent.inherit({
...
-            result;
-
+            rules = this._getValidationRules();
+        const currentResult = this._validationInfo && this._validationInfo.result;
+        if(currentResult && currentResult.status === "pending" && currentResult.value === value) {
+            return currentResult;
+        }
+        let result;
         if(bypass) {
             result = { isValid: true };
```
Comment (DokaRus) [inline js/ui/validator.js:232]: Should it have the `status` field? 
Should we add specific method to get default or specific ValidationResult and use it in places like this or [L254](https://github.com/DevExpress/DevExtreme/pull/9655/files#diff-5f07271123f260b42dd2aaa84e126109L254)?

```js
validationEngine?
prepareValidationResult(result) {
    const defaultResult = {
        isValid: true,
        brokenRule: null,
        brokenRules: null,
        pendingRules: null,
        status: "valid",
        complete: null
    };
    return extend(true, defaultResult, result)
}

**Review by DokaRus — COMMENTED** [review]
Could you add tests for the editor with pending indicator? Consider disposing and repainting the widget while pending

File: js/integration/knockout/validation.js (hunk 2 above)
Comment (NickMitrokhin) [reply js/integration/knockout/validation.js:17]: I added _validationInfo. See [L24](https://github.com/DevExpress/DevExtreme/pull/9655/files#diff-e467507876ea61a57878947822a8ca60R24).

File: js/ui/button.js (hunk 4 above)
Comment (NickMitrokhin) [reply js/ui/button.js]: OK

File: js/ui/editor/editor.js (hunk 7 above)
Comment (NickMitrokhin) [reply js/ui/editor/editor.js:235]: validationErrors is a new option. May not be synchronize after with validationError. 
validationError is going to be deprecated. I do not think that we should remove  L224-L226 now. Only after removing validationError from code after the deprecation.

File: js/ui/editor/editor.js (hunk 8 above)
Comment (NickMitrokhin) [reply js/ui/editor/editor.js]: Makes sense. Will do.

File: js/ui/text_box/ui.text_editor.base.js (hunk 9 above)
Comment (NickMitrokhin) [reply js/ui/text_box/ui.text_editor.base.js]: Makes sense. Will do.

File: js/ui/validation_engine.js (hunk 12 above)
Comment (NickMitrokhin) [reply js/ui/validation_engine.js:8]: Several methods are used from the type module. Do not think that it is better to extract them.

File: js/ui/validator.js (hunk 20 above)
Comment (NickMitrokhin) [reply js/ui/validator.js]: No. It seems to be merged accidentally from previous commits.

File: js/ui/validator.js (hunk 21 above)
Comment (NickMitrokhin) [reply js/ui/validator.js:232]: >Should it have the status field?

Yes. This code for preventing validation when the status is pending and the value is not changed. In this case, the current validation result should be returned.

>Should we add specific method to get default or specific ValidationResult and use it in places like this or L254?

The result is reset only once. I do not think we should create a method for redefining the result object.

**Comment by NickMitrokhin** [comment]
@vconst, I've added tests and fixed dx.all.d.ts.

**Comment by NickMitrokhin** [comment]
@DokaRus 

>Could you add tests for the editor with pending indicator?

Done.

**Review by DokaRus — APPROVED** [review]

**Pull request merged**